### ISSUE TITLE
 calibrate int8 get error of "core dump"

### ISSUE THREAD

**Issue by CoinCheung** [NVIDIA/TensorRT#1120]
## Environment

**TensorRT 7.0.0**: 
**NVIDIA V100**: 
**NVIDIA Driver Version: 450.80**: 
**CUDA Version: 10.2**: 
**CUDNN Version: 7**: 
**Operating System: ubuntu18.04, docker image**: 
**Python Version (if applicable)**: 
**Tensorflow Version (if applicable)**: 
**PyTorch Version (if applicable): 1.8**: 
**Baremetal or Container (if so, version)**: 

## Relevant Files
None

## Steps To Reproduce

Following is how I build the engine: 
```


TrtSharedEnginePtr shared_engine_ptr(ICudaEngine* ptr) {
    return TrtSharedEnginePtr(ptr, TrtDeleter());
}


TrtSharedEnginePtr parse_to_engine(string onnx_pth, bool use_fp16) {
    unsigned int maxBatchSize{1};
    int64_t memory_limit = 1UL << 32; // 1G

    auto builder = TrtUniquePtr<IBuilder>(nvinfer1::createInferBuilder(gLogger));
    CHECK (builder) << "create builder failed\n";

    const auto explicitBatch = 1U << static_cast<uint32_t>(
            nvinfer1::NetworkDefinitionCreationFlag::kEXPLICIT_BATCH);
    auto network = TrtUniquePtr<INetworkDefinition>(
            builder->createNetworkV2(explicitBatch));
    CHECK (network) << "create network failed\n";

    auto config = TrtUniquePtr<IBuilderConfig>(builder->createBuilderConfig());
    CHECK (config) << "create builder config failed\n";

    auto parser = TrtUniquePtr<nvonnxparser::IParser>(nvonnxparser::createParser(*network, gLogger));
    CHECK (parser) << "create parser failed\n";

    int verbosity = (int)nvinfer1::ILogger::Severity::kWARNING;
    bool state = parser->parseFromFile(onnx_pth.c_str(), verbosity);
    CHECK (state) << "parse model failed\n";


    config->setMaxWorkspaceSize(memory_limit);
    if (use_fp16 && builder->platformHasFastFp16()) {
        config->setFlag(nvinfer1::BuilderFlag::kFP16); // fp16
    }

    // TODO: add an option to do this
    builder->setMaxBatchSize(128);
    bool use_int8 = true;
    if (use_int8 && builder->platformHasFastInt8()) {
        config->setFlag(nvinfer1::BuilderFlag::kINT8); // int8 switch on
        // calibrate
        // TODO: here params
        int batchsize = 1;
        int n_cal_batches = 100;
        string data_file = "/root/GEDeepLab_yaogan/datasets/cityscapes/train.txt";
        string data_root = "/root/GEDeepLab_yaogan/datasets/cityscapes";
        string network_name = "deeplab_";
        string input_name = "input_image";
        BatchStream calibrationStream(batchsize, n_cal_batches, data_root, data_file);
        std::unique_ptr<IInt8Calibrator> calibrator;
        // TODO: here name
        calibrator.reset(new Int8EntropyCalibrator2<BatchStream>(
            calibrationStream, 0, network_name.c_str(), input_name.c_str()));
        config->setInt8Calibrator(calibrator.get());
    }
    // TODO: see if use dla

    // change output to int, since the output is argmax
    // int n_layers = network->getNbLayers();
    // auto out_layer = network->getLayer(n_layers - 1);
    // out_layer->setOutputType(0, nvinfer1::DataType::kINT32);
    auto output = network->getOutput(0);
    output->setType(nvinfer1::DataType::kINT32);

    cout << " start to build \n";
    TrtSharedEnginePtr engine = shared_engine_ptr(
            builder->buildEngineWithConfig(*network, *config));
    if (!engine) {
        gLogger.log(nvinfer1::ILogger::Severity::kERROR, "create engine failed");
        return nullptr;
    }

    return engine;
}
```
And following is how I implemented batch_stream and entropy_calibrator, I just modified from the example to write this: 
```
// batch_stream.hpp
#include <string>
#include <sstream>
#include <fstream>
#include <iostream>
#include <vector>
#include <algorithm>
#include <numeric>
#include <opencv2/opencv.hpp>

#include "NvInfer.h"
#include "read_img.hpp"

using nvinfer1::Dims3;
using nvinfer1::Dims4;


class IBatchStream
{
public:
    virtual void reset(int firstBatch) = 0;
    virtual bool next() = 0;
    virtual void skip(int skipCount) = 0;
    virtual float* getBatch() = 0;
    virtual int getBatchesRead() const = 0;
    virtual int getBatchSize() const = 0;
    virtual nvinfer1::Dims4 getDims() const = 0;
};


class BatchStream : public IBatchStream
{
public:
    BatchStream(int batchSize, int maxBatches,
            const std::string& dataFile, const std::string& dataRoot)
        : mBatchSize{batchSize}
        , mMaxBatches{maxBatches}
        , mDims{3, 1024, 2048}
    {
        // readDataFile(locateFile(dataFile, directories));
        readDataFile(dataFile, dataRoot);
        mSampleSize = std::accumulate(
                mDims.d, mDims.d + mDims.nbDims, 1, std::multiplies<int64_t>());
        mData.resize(mSampleSize * mBatchSize);
    }

    void reset(int firstBatch) override
    {
        cout << "call reset() from batchstream" << endl;
        cout << "mBatchCount: " << mBatchCount << endl;
        mBatchCount = firstBatch;
    }

    bool next() override
    {
        cout << "call next()" << endl;
        if (mBatchCount >= mMaxBatches)
        {
            return false;
        }
        ++mBatchCount;
        return true;
    }

    void skip(int skipCount) override
    {
        cout << "call skip()" << endl;
        mBatchCount += skipCount;
    }

    // COIN: seems only need to modify this
    float* getBatch() override
    {
        cout << "call getBatch() from batchstream" << endl;
        int offset = mBatchCount * mBatchSize;
        for (int i{0}; i < mBatchSize; ++i) {
            int ind = offset + i;
            cout << "read path: " << mPaths[ind] << endl;
            read_data(mPaths[ind], &mData[i * mSampleSize], mDims.d[1], mDims.d[2]);
        }
        return mData.data();
    }

    int getBatchesRead() const override
    {
        cout << "call getBatchesRead() from batchstream" << endl;
        return mBatchCount;
    }

    int getBatchSize() const override
    {
        cout << "call getBatchSize() from batchstream" << endl;
        return mBatchSize;
    }

    nvinfer1::Dims4 getDims() const override
    {
        // return Dims4{4, {mBatchSize, mDims.d[0], mDims.d[1], mDims.d[2]}, {}};
        cout << "call getDims() from batchstream" << endl;
        return Dims4{mBatchSize, mDims.d[0], mDims.d[1], mDims.d[2]};
    }

private:
    void readDataFile(const std::string& dataFilePath, const std::string& dataRootPath)
    {
        cout << "call readDataFile() from batchstream" << endl;
        std::ifstream file{dataFilePath.c_str(), std::ios::in};
        std::stringstream ss;
        file >> ss.rdbuf();
        file.close();

        std::string impth;
        int n_imgs = 0;
        while (std::getline(ss, impth)) ++n_imgs;
        ss.clear(); ss.seekg(0, std::ios::beg);

        mPaths.resize(n_imgs);
        for (int i{0}; i < n_imgs; ++i) {
            std::getline(ss, impth, ',');
            mPaths[i] = dataRootPath + "/" + impth;
            std::getline(ss, impth);
        }
        if (mMaxBatches < 0) {
            mMaxBatches =  n_imgs / mBatchSize;
        }
    }


    int mBatchSize{0};
    int mBatchCount{0}; //!< The batch that will be read on the next invocation of next()
    int mMaxBatches{0};
    Dims3 mDims{};
    std::vector<string> mPaths;
    std::vector<float> mData;
    int mSampleSize{0};
};


// entropy_calibrator.hpp
#include <algorithm>
#include <numeric>
#include <iterator>
#include "NvInfer.h"
#include <glog/logging.h>

//! \class EntropyCalibratorImpl
//!
//! \brief Implements common functionality for Entropy calibrators.
//!
template <typename TBatchStream>
class EntropyCalibratorImpl
{
public:
    EntropyCalibratorImpl(
        TBatchStream stream, int firstBatch, std::string networkName, const char* inputBlobName, bool readCache = true)
        : mStream{stream}
        , mCalibrationTableName("CalibrationTable" + networkName)
        , mInputBlobName(inputBlobName)
        , mReadCache(readCache)
    {
        nvinfer1::Dims dims = mStream.getDims();
        mInputCount = std::accumulate(
                dims.d, dims.d + dims.nbDims, 1, std::multiplies<int64_t>());

        // cudaError_t status = cudaMalloc(&mDeviceInput, mInputCount * sizeof(float));
        // cout << "cudaMalloc status: " << status << endl;
        CHECK(!cudaMalloc(&mDeviceInput, mInputCount * sizeof(float)));
        mStream.reset(firstBatch);
    }

    virtual ~EntropyCalibratorImpl()
    {
        CHECK(!cudaFree(mDeviceInput));
    }

    int getBatchSize() const
    {
        cout << "call getBatchSize()" << endl;
        return mStream.getBatchSize();
    }

    bool getBatch(void* bindings[], const char* names[], int nbBindings)
    {
        cout << "call getBatch \n";
        if (!mStream.next())
        {
            return false;
        }
        CHECK(!cudaMemcpy(mDeviceInput, mStream.getBatch(), mInputCount * sizeof(float), cudaMemcpyHostToDevice));
        assert(!strcmp(names[0], mInputBlobName));
        bindings[0] = mDeviceInput;
        return true;
    }

    const void* readCalibrationCache(size_t& length)
    {
        cout << "call readCalibrationCache()" << endl;

        mCalibrationCache.clear();
        std::ifstream input(mCalibrationTableName, std::ios::binary);
        input >> std::noskipws;
        if (mReadCache && input.good())
        {
            std::copy(std::istream_iterator<char>(input), std::istream_iterator<char>(),
                std::back_inserter(mCalibrationCache));
        }
        length = mCalibrationCache.size();
        return length ? mCalibrationCache.data() : nullptr;
    }

    void writeCalibrationCache(const void* cache, size_t length)
    {
        cout << "call writeCalibrationCache()" << endl;
        std::ofstream output(mCalibrationTableName, std::ios::binary);
        output.write(reinterpret_cast<const char*>(cache), length);
    }

private:
    TBatchStream mStream;
    size_t mInputCount;
    std::string mCalibrationTableName;
    const char* mInputBlobName;
    bool mReadCache{true};
    void* mDeviceInput{nullptr};
    std::vector<char> mCalibrationCache;
};

//! \class Int8EntropyCalibrator2
//!
//! \brief Implements Entropy calibrator 2.
//!  CalibrationAlgoType is kENTROPY_CALIBRATION_2.
//!
template <typename TBatchStream>
class Int8EntropyCalibrator2 : public nvinfer1::IInt8EntropyCalibrator2
{
public:
    Int8EntropyCalibrator2(
        TBatchStream stream, int firstBatch, const char* networkName, const char* inputBlobName, bool readCache = true)
        : mImpl(stream, firstBatch, networkName, inputBlobName, readCache)
    {
    }

    int getBatchSize() const override
    {
        cout << "call getBatchSize() from calibrator" << endl;
        return mImpl.getBatchSize();
    }

    bool getBatch(void* bindings[], const char* names[], int nbBindings) override
    {
        cout << "call getBatch() from calibrator" << endl;
        return mImpl.getBatch(bindings, names, nbBindings);
    }

    const void* readCalibrationCache(size_t& length) override
    {
        cout << "call readCalibrationCache() from calibrator" << endl;
        return mImpl.readCalibrationCache(length);
    }

    void writeCalibrationCache(const void* cache, size_t length) override
    {
        cout << "call writeCalibrationCache() from calibrator" << endl;
        mImpl.writeCalibrationCache(cache, length);
    }

private:
    EntropyCalibratorImpl<TBatchStream> mImpl;
};
```

I saw not special error messages apart from a line of `Segmentation fault (core dumped)`. Would you tell me how could I make my code work please ?

**Comment by ttyio** [comment]
Hello @CoinCheung , could you check your onnx file using visualize tool like netron? what the shape of network input? thanks!

**Comment by CoinCheung** [comment]
Hi,

I uploaded my model to google drive, the linke is [here](https://drive.google.com/file/d/1o21xUoDgbLHa51DLCkKFabUWZiBmSa9x/view?usp=sharing). I can compile it with fp32 mode and fp16 mode, but I cannot compile it with int8 calibration.

The input shape is (1, 3, 1024, 2048), I exported it from pytorch:
```
dummy_input = torch.randn(1, 3, 1024, 2048)
input_names = ['input_image']
output_names = ['preds',]

torch.onnx.export(net, dummy_input, args.out_pth,
    input_names=input_names, output_names=output_names,
    verbose=False, opset_version=10)
```

When I run it with `cuda-memcheck ./main ....`. I got the error message of :
```
 The application may have hit an error when dereferencing Unified Memory from the host. Please rerun the application under cuda-gdb or Nsight Eclipse Edition to catch host side errors.
```

Hope these information is useful.

**Comment by ttyio** [comment]
Hello @CoinCheung , could you help me to do another test, let's move the line

       std::unique_ptr<IInt8Calibrator> calibrator;

to the global scope, make sure the `calibrator` life cycle the same/longer than API `buildEngineWithConfig`? thanks!

**Comment by CoinCheung** [comment]
I can make it work now, the problem is as you assumed that I did not define calibrator outside of the `if(...)` domain.  Thanks a lot for your help !!!

**Comment by CoinCheung** [comment]
Sorry for bothering you again. What is the purpose of batch size for calibration ? Is it only about the speed and memory usage of calibration or is it relavant to the accuracy of the calibrated int8 model. Should I set it to be same as the batchsize that I used to export the onnx model(also same as the batch I run inference in reality) ?

**Comment by ttyio** [comment]
Hello @CoinCheung , your onnx is export with static batch size, so you have to use calibration batch the same as the onnx file.

If you export the onnx with dynamic batch, then you can run calibration using different batch size. The batch size have impact on the accuracy result, because we need collect histogram for calibration, and the histogram is updated after each batch has been inferenced. The histogram expand process is slightly different when calibrate with different batch size. However, the batch size is not as important as other factors like calibration algorithm, calibration data...

**Comment by CoinCheung** [comment]
So if we export onnx model with dynamic batch size, we can use different batch sizes when we compile the engine, calibrate int8 model, and run inference with the engine. While if we export the onnx model with static batch size(like I did), we can only use the same batch size for calibration and inference(batch size should be same as exported on everything). Am I right on this ?

**Comment by ttyio** [comment]
Yes , you are right.

**Comment by CoinCheung** [comment]
Thanks for telling me this !!!
